File: src/test/java/testingil/webinar/tddapi/s3/api_tests_calling_real_controller/NavigatorAPITests.java
package testingil.webinar.tddapi.s3.api_tests_calling_real_controller;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.Matchers.lessThan;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.MvcResult;
import org.springframework.test.web.servlet.setup.MockMvcBuilders;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;

import testingil.webinar.tddapi.s3.Distance;
import testingil.webinar.tddapi.s3.Location;
import testingil.webinar.tddapi.s3.NavigationController;

@SpringBootTest
class NavigatorAPITests {

	@Test
	void navigate_to_same_location_distance_is_zero() throws Exception {
		Location location = new Location("New York");
		setStartPoint(location);
		setDestination(location);
		assertThat(distanceTo(location), is(0));
	}

	@Test
	void navigate_and_drive_to_another_location_distance_is_reduced() throws Exception {
		Location initialLocation = new Location("New York");
		setStartPoint(initialLocation);
		Location destination = new Location("Los Angeles");;
		setDestination(destination);
		int initialDistance = distanceTo(destination);
		Location midLocation = new Location("Dallas");
		driveTo(midLocation); 
		assertThat(distanceTo(destination), is(lessThan(initialDistance)));
	}

	MockMvc mockMvc;
	@BeforeEach
	public void setup() {
		this.mockMvc = MockMvcBuilders.standaloneSetup(new NavigationController()).build();
	}
	
	private void setDestination(Location location) throws JsonProcessingException, Exception {
		mockMvc.perform(post("/navigator/destination")
				.content(toJson(location)))
        .andExpect(status().isOk());
	}

	private void setStartPoint(Location location) throws Exception {
		mockMvc.perform(post("/navigator/startpoint")
				.content(toJson(location)))
        .andExpect(status().isOk());
	}

	private void driveTo(Location midLocation) throws Exception {
		setStartPoint(midLocation);
	}
	
	private int distanceTo(Location location) throws JsonProcessingException, Exception {
		MvcResult result = mockMvc.perform(get("/navigator/distance")
				.content(toJson(location)))
        .andExpect(status().isOk())
        .andReturn();
		
		String content = result.getResponse().getContentAsString();
		Distance distance = distanceFromJson(content);
		return distance.inKm();
	}

	private Distance distanceFromJson(String content) throws Exception {
		ObjectMapper objectMapper = new ObjectMapper();
		return objectMapper.readValue(content, Distance.class);
	}

	private String toJson(Location loc) throws Exception {
		ObjectMapper objectMapper = new ObjectMapper();
		return objectMapper.writeValueAsString(loc);
	}

}

File: Python/s03_api_tests_calling_controller/navigator_api_tests.py
import json
import unittest

from hamcrest import assert_that, equal_to, less_than

from s03_api_tests_calling_controller.navigation_controller import app
from s03_api_tests_calling_controller.distance import Distance
from s03_api_tests_calling_controller.location import Location


class NavigatorAPITests(unittest.TestCase):
    def test_navigate_to_same_location_distance_is_zero(self):
        location = Location("New York")
        self.set_start_point(location)
        self.set_destination(location)
        assert_that(self.distance_to_destination(), equal_to(0))

    def test_navigate_and_drive_to_another_location_distance_is_reduced(self):
        initialLocation = Location("New York")
        self.set_start_point(initialLocation)

        destination = Location("Los Angeles")
        self.set_destination(destination)

        initialDistance = self.distance_to_destination()

        midLocation = Location("Dallas")
        self.drive_to(midLocation)

        assert_that(self.distance_to_destination(), less_than(initialDistance))

    def set_start_point(self, location):
        response = self.app.post("/nav/startpoint",
                                 content_type='application/json',
                                 data=json.dumps(location.__dict__))
        self.assertEqual(200, response.status_code)

    def set_destination(self, location):
        response = self.app.post("/nav/destination",
                                 content_type='application/json',
                                 data=json.dumps(location.__dict__))
        self.assertEqual(200, response.status_code)

    def distance_to_destination(self):
        response = self.app.get("/nav/distance")
        self.assertEqual(200, response.status_code)
        distance = Distance(response.get_json()['miles'])
        return distance.inKm()

    def drive_to(self, location):
        self.set_start_point(location)

    def setUp(self):
        self.app = app.test_client(self)


if __name__ == '__main__':
    unittest.main()
